crates/ra_ide_api/src/completion/completion_item.rs

```diff
@@ -11,15 +15,29 @@ pub struct CompletionItem {
     /// completion.
     completion_kind: CompletionKind,
     label: String,
+    kind: Option<CompletionItemKind>,
     detail: Option<String>,
     lookup: Option<String>,
-    snippet: Option<String>,
-    kind: Option<CompletionItemKind>,
-}
-
-pub enum InsertText {
-    PlainText { text: String },
-    Snippet { text: String },
+    /// The format of the insert text. The format applies to both the `insert_text` property
+    /// and the `insert` property of a provided `text_edit`.
+    insert_text_format: InsertTextFormat,
+    /// An edit which is applied to a document when selecting this completion. When an edit is
+    /// provided the value of `insert_text` is ignored.
+    ///
+    /// *Note:* The range of the edit must be a single line range and it must contain the position
+    /// at which completion has been requested.
+    ///
+    /// *Note:* If sending a range that overlaps a string, the string should match the relevant
+    /// part of the replacement text, or be filtered out.
+    text_edit: Option<AtomTextEdit>,
+    /// An optional array of additional text edits that are applied when
+    /// selecting this completion. Edits must not overlap (including the same insert position)
+    /// with the main edit nor with themselves.
+    ///
+    /// Additional text edits should be used to change text unrelated to the current cursor position
+    /// (for example adding an import statement at the top of the file if the completion item will
+    /// insert an unqualified type).
+    additional_text_edits: Option<TextEdit>,
```

Comment:
I feel uneasy about this change of the fields. 

They mirror the field of the corresponding LSP type, but I am not sure that its design is actually good.

In particular, it seems to me that the separation between `text_edit` and `additional_text_edits` is just a backwards-compatibility issue.

Can we start with a more minimal change for now? In particular, I think just adding a `replace_range: TextRange` to the `CompletionItem` should be enough to produce a text edit?

That is, the only logical bit of information that we are missing is which identifier gets replaced by completion, and it seems totally reasonable that we should add it. Factoring the text edit into "primary" and "secondary" ones does not seem right to me: what is the purpose of the factoring, besides the "lsp protocol works this way"?

Reply:
`replace_range` is not enough to product `text_edit` and `additional_text_edits`.

```
Struct A {
    first_method: usize,
    second_method: usize
}

a = A {xxx}
a.first<|>
```

converts to  `replace range {2, 7}`

```
a.whi<|>
```

should converts to 

```
replace range  {2, 5}
delete range {0, 2}
```

Reply:
That's specific to postfix completion, right? 

That's interesting! Abstractly, I'd expect `CompletionItem` to contain two things:

```rust
// The range of the identifier that is being completed. 
// This is purely advisory and can be used, for example, to highlight this range in the editor.
// Clients are expected to ignore this field.
source_range: TextRange, 
// The change to apply to the document when the edit is selected. It 
// typically replaced identifer at `source_range`, but it can contain arbitrary other edits as well.
// The edits themself may contain snippet placeholders like `$0`
edit: SourceFileEdit
```

... after digging through LSP's git blame for half an hour:

OK, I've found https://github.com/Microsoft/language-server-protocol/issues/92, will leave a comment there. 

Not sure what's the way forward here for postfix completion. Could we keep an isolated ugly hack specifically for them?

Reply:
i will try this way tomorrow. i guess it will work. No need for ugly hack.

Reply:
https://github.com/Microsoft/language-server-protocol/issues/92#issuecomment-455797531